sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -704,21 +772,40 @@ public void populateDisplayData(DisplayData.Builder builder) {
                           getQuery(),
                           getParameterSetter(),
                           getRowMapper(),
-                          getFetchSize())))
+                          getFetchSize(),
+                          getBeamSchemaRowMapper(),
+                          getWithBeamSchema() != null && getWithBeamSchema(),
+                          schema)))
               .setCoder(getCoder());
-
       if (getOutputParallelization()) {
         output = output.apply(new Reparallelize<>());
       }
 
+      if (getWithBeamSchema() != null && getWithBeamSchema()) {
+        output = output.setRowSchema(schema);
+      }
+
       return output;
     }
 
+    private Schema generateSchemaFromQuery() {
+      try {
+        Connection connection = getDataSourceProviderFn().apply(null).getConnection();
```

Comment:
Connection and prepared statement is leaking here.

Reply:
Done